File: backend/src/services/user.service.ts
import { query } from '../db/client';
import { User, UserRole } from '../types';

export class UserService {
  async createUser(
    cognitoSub: string,
    orgId: string,
    email: string,
    role: UserRole,
    supervisorId?: string
  ): Promise<User> {
    const result = await query(
      `INSERT INTO users (cognito_sub, org_id, email, role, supervisor_id)
       VALUES ($1, $2, $3, $4, $5)
       RETURNING *`,
      [cognitoSub, orgId, email, role, supervisorId || null]
    );
    return result.rows[0];
  }

  async getUserByCognitoSub(cognitoSub: string): Promise<User | null> {
    const result = await query(
      'SELECT * FROM users WHERE cognito_sub = $1',
      [cognitoSub]
    );
    return result.rows[0] || null;
  }

  async getUserById(id: string): Promise<User | null> {
    const result = await query(
      'SELECT * FROM users WHERE id = $1',
      [id]
    );
    return result.rows[0] || null;
  }

  async listUsersByOrganization(orgId: string): Promise<User[]> {
    const result = await query(
      'SELECT * FROM users WHERE org_id = $1 ORDER BY created_at DESC',
      [orgId]
    );
    return result.rows;
  }

  async getSubordinates(supervisorId: string): Promise<User[]> {
    const result = await query(
      'SELECT * FROM users WHERE supervisor_id = $1',
      [supervisorId]
    );
    return result.rows;
  }

  async updateUser(
    id: string,
    updates: Partial<Pick<User, 'email' | 'role' | 'supervisor_id'>>
  ): Promise<User | null> {
    const fields = [];
    const values = [];
    let paramCount = 1;

    if (updates.email) {
      fields.push(`email = $${paramCount++}`);
      values.push(updates.email);
    }
    if (updates.role) {
      fields.push(`role = $${paramCount++}`);
      values.push(updates.role);
    }
    if (updates.supervisor_id !== undefined) {
      fields.push(`supervisor_id = $${paramCount++}`);
      values.push(updates.supervisor_id);
    }

    if (fields.length === 0) {
      return this.getUserById(id);
    }

    values.push(id);
    const result = await query(
      `UPDATE users 
       SET ${fields.join(', ')}
       WHERE id = $${paramCount}
       RETURNING *`,
      values
    );

    return result.rows[0] || null;
  }

  async deleteUser(id: string): Promise<boolean> {
    const result = await query(
      'DELETE FROM users WHERE id = $1',
      [id]
    );
    return (result.rowCount ?? 0) > 0;
  }
}

File: backend/app/services/user_service.py
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from uuid import UUID
from app.models import User
from app.schemas import UserCreate, UserUpdate


class UserService:
    @staticmethod
    async def create(db: AsyncSession, user_data: UserCreate) -> User:
        """Create a new user"""
        user = User(**user_data.model_dump())
        db.add(user)
        await db.commit()
        await db.refresh(user)
        return user
    
    @staticmethod
    async def get_by_id(db: AsyncSession, user_id: UUID) -> User | None:
        """Get user by ID"""
        result = await db.execute(
            select(User).where(User.id == user_id)
        )
        return result.scalar_one_or_none()
    
    @staticmethod
    async def get_by_cognito_sub(db: AsyncSession, cognito_sub: str) -> User | None:
        """Get user by Cognito sub"""
        result = await db.execute(
            select(User).where(User.cognito_sub == cognito_sub)
        )
        return result.scalar_one_or_none()
    
    @staticmethod
    async def list_by_organization(db: AsyncSession, org_id: UUID) -> list[User]:
        """List users in an organization"""
        result = await db.execute(
            select(User)
            .where(User.org_id == org_id)
            .order_by(User.created_at.desc())
        )
        return list(result.scalars().all())
    
    @staticmethod
    async def get_subordinates(db: AsyncSession, supervisor_id: UUID) -> list[User]:
        """Get subordinates of a supervisor"""
        result = await db.execute(
            select(User).where(User.supervisor_id == supervisor_id)
        )
        return list(result.scalars().all())
    
    @staticmethod
    async def update(
        db: AsyncSession,
        user_id: UUID,
        user_data: UserUpdate
    ) -> User | None:
        """Update user"""
        user = await UserService.get_by_id(db, user_id)
        if not user:
            return None
        
        update_data = user_data.model_dump(exclude_unset=True)
        for field, value in update_data.items():
            setattr(user, field, value)
        
        await db.commit()
        await db.refresh(user)
        return user
    
    @staticmethod
    async def delete(db: AsyncSession, user_id: UUID) -> bool:
        """Delete user"""
        user = await UserService.get_by_id(db, user_id)
        if not user:
            return False
        
        await db.delete(user)
        await db.commit()
        return True
